BUG: Fix thinko in assert_deprecated()

Daetalus opened the pull request:
Hi, I am try to let NumPy run in [Pyston](https://github.com/dropbox/pyston). I may need to patch NumPy too. 

The first stage is try to pass NumPy tests in Pyston. Then move to SciPy and performance issue. The ultimate goal is run Scikit-learn in Pyston.

I am in numpy-discussion list. Feel free to contact me and discuss the numpy-pyston support on it.

njsmith commented:
Huh. I wonder how that ever worked on CPython.

(You might want to file a bug on pyston too with a small test case -- in this case it looks like a harmless patch to include in numpy, but in the long run they'll probably want to know about these weird corner cases so that the next project won't need special patching.)

njsmith commented:
Or did this code even work before? If not and this is a real bug fix, then we should also add a test case to make sure it stays fixed.

Daetalus commented:
I think due to NumPy tests is correct, so the code was never been run before.

seberg commented:
How easy is it to test the test reporting, especially as this is a bit hidden away and not part of the test suit?

njsmith commented:
It looks like something like
```
class TestTestDeprecated(_TestDeprecated):
    def test_assert_deprecated(self):
        assert_raises(AssertionError, assert_deprecated, lambda: None)
```
would do it...

seberg commented:
Yeah, you are right. @Daetalus, just to be clear, could you add the test Nathaniel suggested? After that, one of us will merge this.

Daetalus commented:
Hi @seberg  and @njsmith , thanks for your help! The commit updated according your comments. Please review it if you have time.

njsmith commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -400,5 +400,13 @@ def test_fortran_contiguous(self):
         self.assert_deprecated(np.ones((2,2)).T.view, args=(np.int8,))
 
 
+class TestTestDeprecated(_DeprecationTestCase):
```
Ah, if you do this the way you do below (which looks find to me) then you don't need to inherit from `_DeprecationTestCase`.

njsmith commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -400,5 +400,13 @@ def test_fortran_contiguous(self):
+        test_case_instance.setUp()
+        assert_raises(AssertionError,
+                      test_case_instance.assert_deprecated,
+                      lambda: None)
```
I guess we should probably go the extra step and also add
```
def f():
    warnings.warn("foo", category=DeprecationWarning)
test_case_instance.assert_deprecated(f)
```

njsmith commented:
Otherwise looks good to me

Daetalus committed:
BUG: Fix thinko in assert_deprecated()

assert_deprecated() was recently reworked for stylistic changes (in
0aa32608 "STY: Minor style cleanups in tests and C code.") but made a
thinko - `lst` is already a list of warnings, so we don't need to put
that into [] braces when preparing assertion text. If we do the
reporting breaks:

    In [1]: msg = "4 warnings found but 3 expected."
    In [2]: lst = ['CategoryA', 'CategoryB', 'CategoryC']
    In [3]: n.join([msg] + [lst])
    TypeError   Traceback (most recent call last)
    ----> 1 n.join([msg] + [lst])

    TypeError: sequence item 1: expected string, list found

Fix it.

Cc: Charles Harris <charlesr.harris@gmail.com>

Daetalus commented:
@njsmith Thanks for feedback! I updated the commit, this is the inter diff:

```
diff --git a/numpy/core/tests/test_deprecations.py b/numpy/core/tests/test_deprecations.py
index 49400b9..b0759fd 100644
--- a/numpy/core/tests/test_deprecations.py
+++ b/numpy/core/tests/test_deprecations.py
@@ -400,7 +400,7 @@ class TestNonCContiguousViewDeprecation(_DeprecationTestCase):
         self.assert_deprecated(np.ones((2,2)).T.view, args=(np.int8,))
 
 
-class TestTestDeprecated(_DeprecationTestCase):
+class TestTestDeprecated():
     def test_assert_deprecated(self):
         test_case_instance = _DeprecationTestCase()
         test_case_instance.setUp()
@@ -408,5 +408,10 @@ class TestTestDeprecated(_DeprecationTestCase):
                       test_case_instance.assert_deprecated,
                       lambda: None)
 
+        def foo():
+            warnings.warn("foo", category=DeprecationWarning)
+
+        test_case_instance.assert_deprecated(foo)
+
 if __name__ == "__main__":
     run_module_suite()
(END)
```

njsmith commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -400,5 +400,18 @@ def test_fortran_contiguous(self):
         self.assert_deprecated(np.ones((2,2)).T.view, args=(np.int8,))
 
 
+class TestTestDeprecated():
```
One more tiny tweak, sorry :-). As an issue of style and avoiding weird/surprising bugs, every class should always inherit from `object`, to avoid invoking the weird old-style class infrastructure on python 2.

Let me know when you've had a chance to fix this and I'll merge :-)

Daetalus commented:
@njsmith Updated, my fault, thanks for point out!

njsmith commented:
Great, thanks for bearing with us!
